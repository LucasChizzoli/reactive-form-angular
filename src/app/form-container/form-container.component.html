<div class="form-container-wrapper">
  <h3 class="form-container-title">{{formTitle}}</h3>
  <ng-content></ng-content>
</div>
