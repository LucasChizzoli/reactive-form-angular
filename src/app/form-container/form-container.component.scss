.form-container-wrapper {
  min-width: 400px;

  .form-container-title {
    font-size: 2rem;
    text-align: center;
    margin: 0;
  }
}
