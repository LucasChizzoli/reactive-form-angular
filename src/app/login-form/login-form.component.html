<div class="login-wrapper">
  <h3 class="login-form-title">LOGIN</h3>
  <form class="login-form" [formGroup]="loginForm">
    <div *ngFor="let control of LOGIN_FIELDS">
      <label>
        <span class="form-label">{{control.name | titlecase}}</span>
        <input [type]="getType(control)" class="form-input" [formControlName]="control.name" required>
      </label>
      <div
        *ngIf="loginForm.get(control.name)?.invalid &&
        (loginForm.get(control.name)?.dirty || loginForm.get(control.name)?.touched)">
        <p class="form-error" *ngIf="loginForm.get(control.name)?.errors">
          {{getErrorMessage(control)}}
        </p>
      </div>
    </div>
    <div class="actions-wrapper">
      <button
        class="login-button"
        type="button"
        [disabled]="!loginForm.valid"
        (click)="login()">
        LOGIN
      </button>
    </div>
  </form>
  <div class="login-footer">
    <span>Don't have an account? <a href="#" class="form-link">Register now</a></span>
  </div>
</div>
