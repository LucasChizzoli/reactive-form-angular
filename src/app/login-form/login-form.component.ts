import { Component, EventEmitter } from '@angular/core';
import {
  AbstractControlOptions,
  FormControl,
  FormGroup,
  ValidatorFn,
  Validators,
} from '@angular/forms';
import { LoginModel } from '../models/form-models';

export interface FormField {
  name: string;
  validators: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null;
}

@Component({
  selector: 'app-login-form',
  templateUrl: './login-form.component.html',
  styleUrls: ['./login-form.component.scss']
})
export class LoginFormComponent {
  loginForm: FormGroup;
  loginSubmit = new EventEmitter<LoginModel>();

  // Add, edit or remove the items in the LOGIN_FIELDS array to configure the form
  LOGIN_FIELDS: FormField[] = [
    {
      name: 'email',
      validators: [Validators.email, Validators.required]
    },
    {
      name: 'password',
      validators: [Validators.required, Validators.minLength(8)]
    },
  ];

  constructor() {
    this.loginForm = new FormGroup({});
    this.LOGIN_FIELDS.map(loginField => {
      this.loginForm.addControl(loginField.name, new FormControl('', loginField.validators));
    });
  }

  login(): void {
    if (this.loginForm.invalid) {
      return;
    }
    this.loginSubmit.emit(this.loginForm.value);
  }

  getType(control: FormField): string {
    if (control.name.toLowerCase() === 'email') {
      return 'email';
    }
    if (control.name.toLowerCase() === 'password') {
      return 'password';
    }
    return 'text';
  }

  getErrorMessage(control: FormField): string {
    const controlErrors = this.loginForm.get(control.name)?.errors;
    const controlName = control.name;
    if (!controlErrors) {
      return '';
    }
    if (controlErrors.required) {
      return `${controlName} is required.`;
    }
    if (controlErrors.email) {
      return `email is not valid.`;
    }
    if (controlErrors.minlength) {
      return `${controlName} must be at least ${controlErrors.minlength.requiredLength } long.`;
    }
    if (controlErrors.maxlength) {
      return `${controlName} must be ${controlErrors.maxlength.requiredLength } long.`;
    }
    return '';
  }

}
