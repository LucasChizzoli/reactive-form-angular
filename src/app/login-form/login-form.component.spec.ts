import { ComponentFixture, TestBed } from '@angular/core/testing';
import { LoginFormComponent } from './login-form.component';
import { FormsModule, ReactiveFormsModule, Validators } from '@angular/forms';
import { FormContainerComponent } from '../form-container/form-container.component';
import { FormManagerComponent } from '../form-manager.component';

describe('LoginFormComponent', () => {
  let component: LoginFormComponent;
  let fixture: ComponentFixture<LoginFormComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        FormsModule,
        ReactiveFormsModule
      ],
      declarations: [ LoginFormComponent, FormContainerComponent, FormManagerComponent  ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(LoginFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should have email and password', () => {
    component.fields = [
      {
        name: 'email',
        validators: [Validators.email, Validators.required]
      },
      {
        name: 'password',
        validators: [Validators.required, Validators.minLength(8)]
      },
    ];
    fixture.detectChanges();
    expect(component).toMatchSnapshot();
  });
});
