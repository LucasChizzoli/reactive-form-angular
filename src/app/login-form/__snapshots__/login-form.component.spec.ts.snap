// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginFormComponent should have email and password 1`] = `
LoginFormComponent {
  "LOGIN_FIELDS": Array [
    Object {
      "name": "email",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
    Object {
      "name": "password",
      "validators": Array [
        [Function],
        [Function],
      ],
    },
  ],
  "loginForm": FormGroup {
    "_composedAsyncValidatorFn": null,
    "_composedValidatorFn": null,
    "_hasOwnPendingAsyncValidator": false,
    "_onCollectionChange": [Function],
    "_onDisabledChange": Array [],
    "_parent": null,
    "_rawAsyncValidators": null,
    "_rawValidators": null,
    "controls": Object {
      "email": FormControl {
        "_composedAsyncValidatorFn": null,
        "_composedValidatorFn": [Function],
        "_hasOwnPendingAsyncValidator": false,
        "_onChange": Array [
          [Function],
          [Function],
        ],
        "_onCollectionChange": [Function],
        "_onDisabledChange": Array [
          [Function],
          [Function],
        ],
        "_parent": [Circular],
        "_pendingValue": "",
        "_rawAsyncValidators": null,
        "_rawValidators": Array [
          [Function],
          [Function],
          [Function],
          [Function],
        ],
        "errors": Object {
          "required": true,
        },
        "pristine": true,
        "status": "INVALID",
        "statusChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "touched": false,
        "value": "",
        "valueChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
      "password": FormControl {
        "_composedAsyncValidatorFn": null,
        "_composedValidatorFn": [Function],
        "_hasOwnPendingAsyncValidator": false,
        "_onChange": Array [
          [Function],
          [Function],
        ],
        "_onCollectionChange": [Function],
        "_onDisabledChange": Array [
          [Function],
          [Function],
        ],
        "_parent": [Circular],
        "_pendingValue": "",
        "_rawAsyncValidators": null,
        "_rawValidators": Array [
          [Function],
          [Function],
          [Function],
          [Function],
        ],
        "errors": Object {
          "required": true,
        },
        "pristine": true,
        "status": "INVALID",
        "statusChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "touched": false,
        "value": "",
        "valueChanges": EventEmitter_ {
          "__isAsync": false,
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
      },
    },
    "errors": null,
    "pristine": true,
    "status": "INVALID",
    "statusChanges": EventEmitter_ {
      "__isAsync": false,
      "_isScalar": false,
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "touched": false,
    "value": Object {
      "email": "",
      "password": "",
    },
    "valueChanges": EventEmitter_ {
      "__isAsync": false,
      "_isScalar": false,
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
  },
  "loginSubmit": EventEmitter_ {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
}
`;
